### PR TITLE
moving to next version of s3select.

### PR DESCRIPTION
**Pull request by galsalomon66**
- s3tests points to updated version of s3select tests(https://github.com/ceph/s3-tests/pull/472)
- s3select version https://github.com/ceph/s3select/pull/115

Signed-off-by: galsalomon66 <gal.salomon@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: qa/suites/rgw/verify/tasks/s3tests.yaml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 tasks:
 - s3tests:
     client.0:
+      force-branch: fixes_related_to_new_csv_parser
+      git_remote: https://github.com/galsalomon66/
```
Comment (cbodley) [inline qa/suites/rgw/verify/tasks/s3tests.yaml:5]: this is pointing to a branch based on s3tests' `master` branch, which may contain newer tests that don't pass against ceph yet

i rebased your s3tests `fixes_related_to_new_csv_parser` branch on top of `ceph-master` instead and pushed to https://github.com/cbodley/s3-tests/commits/fixes_related_to_new_csv_parser. i also updated this s3tests.yaml to point the `git_remote` at my cbodley repo, and pushed that single commit to https://github.com/cbodley/ceph/commits/wip-qa-rgw-s3select-fixes_related_to_new_csv_parser

i've pushed this PR to ceph-ci for builds at https://shaman.ceph.com/builds/ceph/pr-47848/

once packages are built, i'll schedule the teuthology tests using my updated --suite-branch, ie:
> $ teuthology-suite -s rgw -m smithi -c pr-47848 --suite-repo https://github.com/cbodley/ceph.git --suite-branch wip-qa-rgw-s3select-fixes_related_to_new_csv_parser -p 75

now that this change to s3tests.yaml is in that suite-branch, feel free to remove it from this PR so we don't accidentally merge it

**Review by cbodley — COMMENTED** [review]
the `Unmodified Submodules` check failed with this error:

> please include 's3select submodule' in your commit message, if this change is intentional.

File: qa/suites/rgw/verify/tasks/s3tests.yaml (hunk 1 above)
Comment (galsalomon66) [reply qa/suites/rgw/verify/tasks/s3tests.yaml:5]: ok,
Thanks.

**Comment by cbodley** [comment]
```
2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:======================================================================
2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:FAIL: s3tests_boto3.functional.test_s3select.test_csv_parser
2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:----------------------------------------------------------------------
2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:Traceback (most recent call last):
2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:  File "/home/ubuntu/cephtest/s3-tests-client.0/virtualenv/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:    self.test(*self.arg)
2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:  File "/home/ubuntu/cephtest/s3-tests-client.0/s3tests_boto3/functional/test_s3select.py", line 902, in test_csv_parser
2022-09-02T04:22:12.686 INFO:teuthology.orchestra.run.smithi071.stderr:    s3select_assert_result( res_s3select_alias, 'third="c31,c32,c33"')
2022-09-02T04:22:12.686 INFO:teuthology.orchestra.run.smithi071.stderr:  File "/home/ubuntu/cephtest/s3-tests-client.0/s3tests_boto3/functional/test_s3select.py", line 115, in s3select_assert_result
2022-09-02T04:22:12.686 INFO:teuthology.orchestra.run.smithi071.stderr:    nose.tools.assert_equal(a,b)
2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:AssertionError: 'An error occurred (s3select-ProcessingTim[78 chars]sing' != 'third="c31,c32,c33"'
2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:- An error occurred (s3select-ProcessingTime-Error) when calling the SelectObjectContent operation: failure while csv parsing
2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:+ third="c31,c32,c33"
2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:
2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:-------------------- >> begin captured logging << --------------------
```
ex http://qa-proxy.ceph.com/teuthology/cbodley-2022-09-01_15:28:29-rgw-pr-47848-distro-default-smithi/7005448/teuthology.log

**Comment by cbodley** [comment]
full results in https://pulpito.ceph.com/cbodley-2022-09-01_15:28:29-rgw-pr-47848-distro-default-smithi

**Comment by cbodley** [comment]
```
2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:FAIL: s3tests_boto3.functional.test_s3select.test_csv_parser
```
i think these are just coming from other subsuites like rgw/crypt and rgw/website that didn't change the s3tests branch

in https://github.com/ceph/ceph/pull/47044, @alimaredia added a `qa/rgw/s3tests-branch.yaml` file that gets symlinked into all of these subsuites, so we can change the repo/branch in one place. i'll update my suite-branch to do that

**Comment by cbodley** [comment]
updated the suite-branch with https://github.com/cbodley/ceph/commit/e20bde8244d2a27c6431acd07e9967aaa8f04f01 and rescheduled in https://pulpito.ceph.com/cbodley-2022-09-02_14:33:28-rgw-pr-47848-distro-default-smithi/

**Comment by galsalomon66** [comment]
> ```
> 2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:======================================================================
> 2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:FAIL: s3tests_boto3.functional.test_s3select.test_csv_parser
> 2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:----------------------------------------------------------------------
> 2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:Traceback (most recent call last):
> 2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:  File "/home/ubuntu/cephtest/s3-tests-client.0/virtualenv/lib/python3.8/site-packages/nose/case.py", line 198, in runTest
> 2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:    self.test(*self.arg)
> 2022-09-02T04:22:12.685 INFO:teuthology.orchestra.run.smithi071.stderr:  File "/home/ubuntu/cephtest/s3-tests-client.0/s3tests_boto3/functional/test_s3select.py", line 902, in test_csv_parser
> 2022-09-02T04:22:12.686 INFO:teuthology.orchestra.run.smithi071.stderr:    s3select_assert_result( res_s3select_alias, 'third="c31,c32,c33"')
> 2022-09-02T04:22:12.686 INFO:teuthology.orchestra.run.smithi071.stderr:  File "/home/ubuntu/cephtest/s3-tests-client.0/s3tests_boto3/functional/test_s3select.py", line 115, in s3select_assert_result
> 2022-09-02T04:22:12.686 INFO:teuthology.orchestra.run.smithi071.stderr:    nose.tools.assert_equal(a,b)
> 2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:AssertionError: 'An error occurred (s3select-ProcessingTim[78 chars]sing' != 'third="c31,c32,c33"'
> 2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:- An error occurred (s3select-ProcessingTime-Error) when calling the SelectObjectContent operation: failure while csv parsing
> 2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:+ third="c31,c32,c33"
> 2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:
> 2022-09-02T04:22:12.687 INFO:teuthology.orchestra.run.smithi071.stderr:-------------------- >> begin captured logging << --------------------
> ```
> 
> ex http://qa-proxy.ceph.com/teuthology/cbodley-2022-09-01_15:28:29-rgw-pr-47848-distro-default-smithi/7005448/teuthology.log

I saw that.
there are 31 failures. 
11 are s3select related (same single test, csv-parser test)
the other 20, do not seems related to s3select.

**Comment by cbodley** [comment]
> the other 20, do not seems related to s3select.

agreed

**Comment by galsalomon66** [comment]
> ```
> 2022-09-02T04:22:12.684 INFO:teuthology.orchestra.run.smithi071.stderr:FAIL: s3tests_boto3.functional.test_s3select.test_csv_parser
> ```
> 
> i think these are just coming from other subsuites like rgw/crypt and rgw/website that didn't change the s3tests branch
> 
> in #47044, @alimaredia added a `qa/rgw/s3tests-branch.yaml` file that gets symlinked into all of these subsuites, so we can change the repo/branch in one place. i'll update my suite-branch to do that

It is unclear to me why this impacts a single and the same test (all tests were verified before pushing to teuthology).
i will follow the next run.

**Comment by galsalomon66** [comment]
> updated the suite-branch with [cbodley@e20bde8](https://github.com/cbodley/ceph/commit/e20bde8244d2a27c6431acd07e9967aaa8f04f01) and rescheduled in https://pulpito.ceph.com/cbodley-2022-09-02_14:33:28-rgw-pr-47848-distro-default-smithi/

all s3select-tests had passed  successfully
there are 17 failures, 8 of them are Valgrind issues, that are s3select related (looks like the same issue)

**Comment by galsalomon66** [comment]
> 
 running Valgrind on s3tests(locally),  did not produce any s3select issue.
 it did produce many Invalid read/write on other call-stack.

**Comment by galsalomon66** [comment]
http://pulpito.front.sepia.ceph.com/gsalomon-2022-09-15_21:18:32-rgw:verify-s3select_next_version-distro-default-smithi/
all 24 s3select tests had passed with success.
there are still valgrind reports on exact same call-stack (s3select::to_string)

**Comment by cbodley** [comment]
i was able to reproduce one of the `to_string()` leaks. the [valgrind report](https://gist.github.com/cbodley/a631ce99a43637ca6966f1aa28af502a) has more debug symbols than the teuthology one, but still isn't all that helpful. it's pointing at [this line](https://github.com/ceph/s3select/blob/80572c04/include/s3select_oper.h#L593) of `to_string()`.. somehow the lexical_cast of a double is causing `std::string` to allocate 2511 bytes, which seems excessively long for a number

there was also an [UninitCondition report](https://gist.github.com/cbodley/fef1fd4ee614b2719b085e743360f263) pointing at [this line](https://github.com/ceph/s3select/blob/80572c0443d6f22b28e2cad08e86a971ebd076d6/include/s3select_csv_parser.h#L351) of `CSVParser::next_line()`

i reproduced this by building with WITH_BOOST_VALGRIND=ON and the default tcmalloc ALLOCATOR, and running radosgw with the same valgrind flags that teuthology does:
> ~/ceph/build $ valgrind --tool=memcheck --trace-children=no --child-silent-after-fork=yes '--soname-synonyms=somalloc=\*tcmalloc\*' --num-callers=50 --suppressions=../qa/valgrind.supp --xml=yes --xml-file=out/valgrind.8000.log --time-stamp=yes --vgdb=yes --tool=memcheck --max-threads=1024 bin/radosgw -c ceph.conf --log-file=out/radosgw.8000.log --admin-socket=out/radosgw.8000.asok --pid-file=out/radosgw.8000.pid --rgw_luarocks_location=out/luarocks --debug-rgw=20 --debug-ms=1 -n client.rgw.8000 --rgw_frontends=beast\ port=8000 -f

**Comment by galsalomon66** [comment]
> i was able to reproduce one of the `to_string()` leaks. the [valgrind report](https://gist.github.com/cbodley/a631ce99a43637ca6966f1aa28af502a) has more debug symbols than the teuthology one, but still isn't all that helpful. it's pointing at [this line](https://github.com/ceph/s3select/blob/80572c04/include/s3select_oper.h#L593) of `to_string()`.. somehow the lexical_cast of a double is causing `std::string` to allocate 2511 bytes, which seems excessively long for a number
> 
> there was also an [UninitCondition report](https://gist.github.com/cbodley/fef1fd4ee614b2719b085e743360f263) pointing at [this line](https://github.com/ceph/s3select/blob/80572c0443d6f22b28e2cad08e86a971ebd076d6/include/s3select_csv_parser.h#L351) of `CSVParser::next_line()`
> 
> i reproduced this by building with WITH_BOOST_VALGRIND=ON and the default tcmalloc ALLOCATOR, and running radosgw with the same valgrind flags that teuthology does:
> 
> > ~/ceph/build $ valgrind --tool=memcheck --trace-children=no --child-silent-after-fork=yes '--soname-synonyms=somalloc=*tcmalloc*' --num-callers=50 --suppressions=../qa/valgrind.supp --xml=yes --xml-file=out/valgrind.8000.log --time-stamp=yes --vgdb=yes --tool=memcheck --max-threads=1024 bin/radosgw -c ceph.conf --log-file=out/radosgw.8000.log --admin-socket=out/radosgw.8000.asok --pid-file=out/radosgw.8000.pid --rgw_luarocks_location=out/luarocks --debug-rgw=20 --debug-ms=1 -n client.rgw.8000 --rgw_frontends=beast\ port=8000 -f

Thanks, @cbodley .
yes, the boost::lexical_case is the "immediate suspect".
does the flag `WITH_BOOST_VALGRIND=ON` and allocator=tcmalloc  "enable" Valrgrind reports on such cases?

**Comment by cbodley** [comment]
WITH_BOOST_VALGRIND filters out the valgrind issues we were seeing around coroutines in https://tracker.ceph.com/issues/56500

the  '--soname-synonyms=somalloc=\*tcmalloc\*' argument filters out lots of tcmalloc-specific issues like 'mismatched malloc/free'

**Comment by galsalomon66** [comment]
- Is it possible to reproduce this boost/double leak quite easily with standalone s3select (no radosgw)
- the resolution for boost::lexical_cast is  std::to_string
- after the fix, Valgrind report on another flow
- a temp refactoring is done to get a better view of Valgrind report

**Comment by galsalomon66** [comment]
http://pulpito.front.sepia.ceph.com/gsalomon-2022-09-20_02:29:45-rgw:verify-s3select_next_version-distro-default-smithi/
all 24 s3select-tests had passed with success.
as for failures not seems to relate to s3select.
no valgrind reports.
the resolution(temporary) is to use s3select-allocator for std::string.
the problem:  std::string assignment(on class value) sometimes causes a leakage. investigating.

**Comment by galsalomon66** [comment]
http://pulpito.front.sepia.ceph.com/gsalomon-2022-09-28_23:04:03-rgw:verify-s3select_next_version-distro-default-smithi/

all 24 s3select tests had passed with success.
no Valgrind reports related to s3select. 
no failures related to s3select.

**Comment by alimaredia** [comment]
@galsalomon66 do you have an s3-tests PR we can merge in parallel with this one? We're going to need to remove all of the `force-branch` and `git_repo` overrides before we merge this PR.

**Comment by cbodley** [comment]
@alimaredia would you be willing to coordinate the testing/merging of these branches?

**Comment by cbodley** [comment]
@alimaredia seeing lots of failures there, are you sure you used the right suite-branch?

**Comment by alimaredia** [comment]
This run should have the correct s3-tests branch: https://pulpito.ceph.com/amaredia-2022-10-13_14:09:35-rgw-s3select_next_version-distro-default-smithi/

**Comment by galsalomon66** [comment]
http://pulpito.front.sepia.ceph.com/gsalomon-2022-10-16_07:14:29-rgw:verify-s3select_next_version-distro-default-smithi/
48 tests had passed with success, including s3select-s3-tests

`grep -a s3tests_boto3.functional.test_s3select /ceph/teuthology-archive//gsalomon-2022-10-16_07:14:29-rgw:verify-s3select_next_version-distro-default-smithi//*/teu*log | awk 'BEGIN{FS=":";} {print $NF;}' | sort | uniq -c`

@cbodley @alimaredia 
please review that.

if that is acceptable, it needs to revert `s3tests-yaml`

**Comment by alimaredia** [comment]
I've been running it with the entire rgw subsuite, not just the verify. Let's see if this one is green, once it is we're ready to merge. https://pulpito.ceph.com/amaredia-2022-10-17_15:08:21-rgw-s3select_next_version-distro-default-smithi/

**Comment by galsalomon66** [comment]
> I've been running it with the entire rgw subsuite, not just the verify. Let's see if this one is green, once it is we're ready to merge. https://pulpito.ceph.com/amaredia-2022-10-17_15:08:21-rgw-s3select_next_version-distro-default-smithi/

ok.
we should forget to revert `s3tests-yaml`

**Comment by cbodley** [comment]
@alimaredia's results are showing some known failures on rhel, so i rescheduled without any rhel jobs using this suite-branch: https://github.com/cbodley/ceph/commits/wip-qa-rgw-s3tests-s3select

@galsalomon66 that suite-branch contains your change to `qa/rgw/s3tests-branch.yaml`, so you can remove that commit from this PR

results pending at https://pulpito.ceph.com/cbodley-2022-10-18_14:19:50-rgw-s3select_next_version-distro-default-smithi/

**Comment by cbodley** [comment]
> https://pulpito.ceph.com/cbodley-2022-10-18_14:19:50-rgw-s3select_next_version-distro-default-smithi/

looks like a pass to me! @galsalomon66 can you please remove the yaml part and squash the rest of the commits?

**Comment by cbodley** [comment]
just needs squashing down to a single commit

**Comment by cbodley** [comment]
i believe the 'ceph API tests' failures have been resolved on main, but you'll need to rebase to pass

**Comment by cbodley** [comment]
> i believe the 'ceph API tests' failures have been resolved on main, but you'll need to rebase to pass

oof, i guess i was wrong. it looks like https://tracker.ceph.com/issues/57834 hasn't been fixed yet :(

**Comment by cbodley** [comment]
jenkins test api

**Comment by cbodley** [comment]
jenkins test api

**Review by cbodley — APPROVED** [review]

**Pull request merged**